File: src/main/java/intricateengineers/intricatemachinery/api/module/MachineryFrame.java
/*
 * Copyright (c) 2016 IntricateEngineers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package intricateengineers.intricatemachinery.api.module;

import intricateengineers.intricatemachinery.api.client.BakedModelFrame;
import intricateengineers.intricatemachinery.api.client.util.UV;
import intricateengineers.intricatemachinery.common.module.DummyModule;
import intricateengineers.intricatemachinery.common.module.FurnaceModule;
import intricateengineers.intricatemachinery.core.ModInfo;

import mcmultipart.MCMultiPartMod;
import mcmultipart.multipart.INormallyOccludingPart;
import mcmultipart.multipart.Multipart;
import mcmultipart.raytrace.RayTraceUtils;

import net.minecraft.block.properties.IProperty;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.util.math.Vec3d;
import net.minecraftforge.common.property.ExtendedBlockState;
import net.minecraftforge.common.property.IExtendedBlockState;
import net.minecraftforge.common.property.IUnlistedProperty;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.annotation.Nullable;

import static net.minecraft.util.EnumFacing.*;

/**
 * @author topisani
 */
public class MachineryFrame extends Multipart implements INormallyOccludingPart {

    public static final Property PROPERTY = new Property();
    public static final BlockModel MODEL = new Model();
    public static final ResourceLocation NAME = new ResourceLocation(ModInfo.MOD_ID.toLowerCase(), "machinery_frame");
    private final Module[][][] modulePositions = new Module[16][16][16];
    public Set<Map<String, ?>> debugInfo;
    private List<AxisAlignedBB> selectionBoxes = new ArrayList<>();
    private List<Module> modules = new ArrayList<>();

    public MachineryFrame() {
        modules.add(new FurnaceModule(this){{setLocalPos(new Vec3d(3 / 16f, 0, 1 / 16f), (byte) 1);}});
        modules.add(new DummyModule(this){{setLocalPos(new Vec3d(0, 5 / 16f, 1 / 16f), (byte) 3);}});
    }

    public List<Module> getModules() {
        return modules;
    }

    public boolean addModule(Module module) {
        //TODO: check occlusion
        modules.add(module);
        return true;
    }

    @Override
    public ResourceLocation getType() {
        return NAME;
    }

    @Override
    public RayTraceUtils.AdvancedRayTraceResultPart collisionRayTrace(Vec3d start, Vec3d end) {
        if (selectionBoxes.isEmpty()) {
            addSelectionBoxes(selectionBoxes);
        }
        RayTraceUtils.AdvancedRayTraceResult result = RayTraceUtils.collisionRayTrace(getWorld(), getPos(), start, end, selectionBoxes);
        return result == null ? null : new RayTraceUtils.AdvancedRayTraceResultPart(result, this);
    }

    @Nullable
    public Module modleHit(Vec3d start, Vec3d end) {
        Vec3d framePos = new Vec3d(this.getPos().getX(), this.getPos().getY(), this.getPos().getZ());
        start = start.add(framePos);
        end = end.add(framePos);
        for (Module module : this.modules) {
            for (AxisAlignedBB bounds : module.getAABBs()) {
                RayTraceResult rt = bounds.offset(module.posX / 16f, module.posY / 16f, module.posZ / 16f).calculateIntercept(start, end);
                if (rt != null) {
                    return module;
                }
            }
        }
        return null;
    }

    /**
     * Adds the selection boxes used to ray trace this part.
     * Called only once when module is placed
     */
    @Override
    public void addSelectionBoxes(List<AxisAlignedBB> list) {
        MODEL.getBoxes().forEach((box) -> list.add(box.toAABB(0, 0, 0)));
        modules.forEach((module) -> module.addSelectionBoxes(list));
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound tag) {
        super.writeToNBT(tag);

        NBTTagCompound modules = new NBTTagCompound();

        for (int i = 0; i < this.modules.size(); i++) { 
            modules.setTag(String.valueOf(i), this.modules.get(i).serializeNBT());
        }

        tag.setTag("modules", modules);

        return tag;
    }

    @Override
    public void readFromNBT(NBTTagCompound tag) {
        super.readFromNBT(tag);
        
        NBTTagList modules = tag.getTagList("modules", 0);

        for (int i = 0; i < modules.tagCount(); i++) {
            //this.modules.add(modules.get(i).
            //TODO: read module
        }

    }

    @Override
    public void writeUpdatePacket(PacketBuffer buf) {
        // TODO: Write modules to packets
    }

    @Override
    public void readUpdatePacket(PacketBuffer buf) {
        // TODO: Read modules from packets
    }

    @Override
    public BlockStateContainer createBlockState() {
        return new ExtendedBlockState(MCMultiPartMod.multipart, new IProperty[0], new IUnlistedProperty[] {PROPERTY});
    }

    @Override
    public IBlockState getActualState(IBlockState state) {

        return state;
    }

    @Override
    public IBlockState getExtendedState(IBlockState state) {
        return ((IExtendedBlockState) state).withProperty(PROPERTY, this);
    }

    @Override
    public void addOcclusionBoxes(List<AxisAlignedBB> list) {
        // TODO: Add boxes from all modules
        list.add(MODEL.mainBox.toAABB(0, 0, 0));
    }

    public Set<Map<String, ?>> getDebugInfo()
    {
        return debugInfo;
    }

    private static class Property implements IUnlistedProperty<MachineryFrame> {

        @Override
        public String getName() {
            return "machinery_frame";
        }

        @Override
        public boolean isValid(MachineryFrame value) {
            return true;
        }

        @Override
        public Class<MachineryFrame> getType() {
            return MachineryFrame.class;
        }

        @Override
        public String valueToString(MachineryFrame value) {
            return value.toString();
        }
    }

    private static class Model extends BlockModel {

        @Override
        public void init() {
            ResourceLocation frameTexture = new ResourceLocation("minecraft", "blocks/furnace_top");
            // Corners
            addBox(vec(0, 0, 0), vec(1, 16, 1))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(0, 0, 15), vec(1, 16, 16))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(15, 0, 0), vec(16, 16, 1))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(15, 0, 15), vec(16, 16, 16))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            // Horizontals
            addBox(vec(1, 0, 0), vec(15, 1, 1))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(1, 15, 0), vec(15, 16, 1))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(0, 0, 1), vec(1, 1, 15))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(0, 15, 1), vec(1, 16, 15))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(15, 0, 1), vec(16, 1, 15))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(15, 15, 1), vec(16, 16, 15))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(1, 0, 15), vec(15, 1, 16))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
            addBox(vec(1, 15, 15), vec(15, 16, 16))
                .setFace(NORTH, frameTexture, UV.auto(16))
                .setFace(EAST, frameTexture, UV.auto(16))
                .setFace(SOUTH, frameTexture, UV.auto(16))
                .setFace(WEST, frameTexture, UV.auto(16))
                .setFace(UP, frameTexture, UV.auto(16))
                .setFace(DOWN, frameTexture, UV.auto(16));
        }

        @Override
        public BakedModelFrame initBakedModel() {
            return new BakedModelFrame();
        }
    }
}

File: src/main/scala/intricateengineers/intricatemachinery/api/module/MachineryFrame.scala
package intricateengineers.intricatemachinery.api.module

import intricateengineers.intricatemachinery.api.client.BakedModelFrame
import intricateengineers.intricatemachinery.api.client.util.UV
import intricateengineers.intricatemachinery.api.model.{BlockModel, Box, BoxFace}
import intricateengineers.intricatemachinery.api.util.{Cache, IHasDebugInfo, Logger}
import intricateengineers.intricatemachinery.common.util.IMRL
import main.scala.intricateengineers.intricatemachinery.api.module.ModuleCapability
import mcmultipart.MCMultiPartMod
import mcmultipart.multipart.{INormallyOccludingPart, Multipart}
import mcmultipart.raytrace.{PartMOP, RayTraceUtils}
import net.minecraft.block.properties.IProperty
import net.minecraft.block.state.{BlockStateContainer, IBlockState}
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.block.model.BakedQuad
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.{NBTTagCompound, NBTTagList}
import net.minecraft.network.PacketBuffer
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.{AxisAlignedBB, RayTraceResult, Vec3d}
import net.minecraftforge.common.property.{ExtendedBlockState, IExtendedBlockState, IUnlistedProperty}
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

import scala.collection.JavaConversions._
import scala.collection.immutable.ListMap
import scala.collection.mutable
import scala.collection.mutable.ListBuffer

object MachineryFrame {
  val NAME = IMRL("machinery_frame")
}

class MachineryFrame extends Multipart
  with INormallyOccludingPart
  with IHasDebugInfo {

  val quadCache: Cache[java.util.List[BakedQuad]] = Cache(updateQuads)
  val bbCache: Cache[java.util.List[AxisAlignedBB]] = Cache(updateAABBs)

  def moduleHit(start: Vec3d, end: Vec3d): Option[Module] = {
    val framePos: Vec3d = new Vec3d(this.getPos.getX, this.getPos.getY, this.getPos.getZ)

    // Gather all hit vectors (positions of the hit) for every module (in the current MF) our eye can raytrace
    // Add them with the associated Module in a Tuple, so that we know to pick the closest one later
    val modulesHit: Map[Vec3d, Module] =
      modules.flatMap(module =>
        module.bbCache().map(bounds =>
          Option(     // At times like this I wish Forge was in Scala
            bounds.calculateIntercept(start.subtract(framePos), end.subtract(framePos)))
                  .map(rayTraceResult => (rayTraceResult.hitVec, module))
      )
    ).flatten.toMap[Vec3d, Module]

    // Get the module for which the associated hit vector is closest
    modulesHit get modulesHit.keys.minBy(_.distanceTo(start))
  }


  @SideOnly(Side.CLIENT)
  def moduleHitFromEyes(): Option[Module] = {
    val mc = Minecraft.getMinecraft
    val eyes: Vec3d = mc.thePlayer.getPositionEyes(1)

    // 5 is the range that AABBs get highlighted (in blocks)
    moduleHit(eyes, eyes.add(mc.thePlayer.getLookVec.scale(5)))
  }

  def wasUpdated(): Unit = {
    modules.invalidate()
  }

  def updateAABBs(): java.util.List[AxisAlignedBB] = {
    val bbs = ListBuffer[AxisAlignedBB]()
    bbs ++= modules.flatMap(_.bbCache())
    bbs ++= FrameModel.aabbs
  }

  @SideOnly(Side.CLIENT)
  private def updateQuads(): java.util.List[BakedQuad] = {
    val buffer = ListBuffer[BakedQuad]()
    buffer ++= FrameModel.boxes.flatMap(_.quads)
    modules.foreach(_.boxCache.invalidate())
    bbCache.invalidate()
    buffer ++= modules.flatMap(_.boxCache().flatMap(_.quads))
    buffer
  }

  private def breakModule(module: Option[Module]): Unit = {
    if(module.isEmpty) Logger.warn("Couldn't find module")
    module.foreach(m =>
      modules -= m)
  }

  /* ------======================------
               OVERRIDES
     ------======================------ */

  override def getType: ResourceLocation = MachineryFrame.NAME

  override def onRemoved(): Unit = {}

  override def harvest(player: EntityPlayer, hit: PartMOP) = {
    val maybeModule = moduleHitFromEyes()
    maybeModule match {
      case Some(module) =>
        breakModule(maybeModule)
      case None =>
        super.harvest(player, hit)
    }
  }

  override def collisionRayTrace(start: Vec3d, end: Vec3d): RayTraceUtils.AdvancedRayTraceResultPart = {
    val result: RayTraceUtils.AdvancedRayTraceResult = RayTraceUtils.collisionRayTrace(getWorld, getPos, start, end,
      bbCache())
    if (result == null) null
    else new RayTraceUtils.AdvancedRayTraceResultPart(result, this)
  }

  override def updateDebugInfo(): ListMap[String, String] = {
    ListMap[String, String](
      "Modules" -> modules.toList.length.toString
    )
  }

  override def writeToNBT(tag: NBTTagCompound): NBTTagCompound = {
    super.writeToNBT(tag)
    val modules: NBTTagList = new NBTTagList
    for (i <- this.modules.toList().indices) {
      try {
        modules.appendTag(this.modules(i).serializeNBT)
      } catch {
        case e: Exception => Logger.warn("Couldn't write to NBT tag")
      }
    }
    tag.setTag("modules", modules)
    tag
  }

  override def readFromNBT(tag: NBTTagCompound) {
    super.readFromNBT(tag)
    val moduleTag: NBTTagList = tag.getTagList("modules", 10)
    modules.clear()
    for (i <- 0 until moduleTag.tagCount) {
      try {
        val mTag = moduleTag.getCompoundTagAt(i)
        val mType = new ResourceLocation(mTag.getString("module_type"))
        (modules += Modules.createModule(mType)(this)).deserializeNBT(mTag)
      } catch {
        case e: Exception => Logger.warn("Couldn't read from NBT tag")
      }
    }
  }

  override def writeUpdatePacket(buf: PacketBuffer) {
    val tag = new NBTTagCompound
    buf.writeNBTTagCompoundToBuffer(writeToNBT(tag))
  }

  override def readUpdatePacket(buf: PacketBuffer) {
    readFromNBT(buf.readNBTTagCompoundFromBuffer())
  }

  override def createBlockState: BlockStateContainer = {
    new ExtendedBlockState(MCMultiPartMod.multipart, new Array[IProperty[_]](0), Array[IUnlistedProperty[_]]
      (FrameProperty))
  }

  override def getActualState(state: IBlockState): IBlockState = state

  override def getExtendedState(state: IBlockState): IBlockState = {
    state.asInstanceOf[IExtendedBlockState].withProperty(FrameProperty, this)
  }

  override def addOcclusionBoxes(list: java.util.List[AxisAlignedBB]): Unit = addSelectionBoxes(list)

  override def addSelectionBoxes(list: java.util.List[AxisAlignedBB]) {
    list.appendAll(bbCache())
  }

  val modules = new Traversable[Module] {
    private val positions = mutable.Map[(Int, Int, Int), Module]()
    private val capabilities = mutable.Map[(Int, Int, Int), ModuleCapability]()
    private val modules = ListBuffer[Module]()

    def apply(i: Int): Module = {
      modules(i)
    }

    def +=(m: Module): Module = {
      modules += m
      forEachCoord(m, (b, f, x, y, z) =>
        positions((x + m.pos.iX, y + m.pos.iY, z + m.pos.iZ)) = m)
      invalidate()
      m
    }

    def -=(m: Module): Unit = {
      modules -= m
      invalidate()
    }

    def clear(): Unit = {
      modules.clear()
    }

    def invalidate(): Unit = {
      bbCache.invalidate()
      quadCache.invalidate()
      debugInfo.invalidate()
      sendUpdatePacket()
    }

    def forEachCoord(m: Module, f: (Box, BoxFace, Int, Int, Int) => Unit): Unit = {
      for (b <- m.boxCache())
        forEachCoord(b, (face, x, y, z) => f(b, face, x, y, z))
    }

    def forEachCoord(b: Box, f: (BoxFace, Int, Int, Int) => Unit): Unit = {
      for (face <- b.faces) {
        val (from, to) = b.vecs(face)
        for (x <- from.x.toInt until to.x.toInt;
          y <- from.y.toInt until to.y.toInt;
          z <- from.z.toInt until to.z.toInt)
          f(face, x, y, z)
      }
    }

    def forEachAdjacent(b: Box, f: (BoxFace, Int, Int, Int) => Unit): Unit = {
      forEachCoord(b, (face, x, y, z) => {
        val offset = face.side.getDirectionVec
        f(face, x + offset.getX, y + offset.getY, z + offset.getZ)
      })
    }

    def forEachAdjacent(m: Module, f: (Box, BoxFace, Int, Int, Int) => Unit): Unit = {
      for (b <- m.boxCache())
        forEachAdjacent(b, (face, x, y, z) => f(b, face, x, y, z))
    }

    override def foreach[U](f: (Module) => U) = modules.foreach(f)

    override def toList(): List[Module] = modules.toList
  }
}

object FrameProperty extends IUnlistedProperty[MachineryFrame] {
  def getName: String = "machinery_frame"

  def isValid(value: MachineryFrame): Boolean = true

  def getType: Class[MachineryFrame] = classOf[MachineryFrame]

  def valueToString(value: MachineryFrame): String = value.toString
}

object FrameModel extends BlockModel {

  lazy val aabbs = boxes.map(_.aabb)
  @SideOnly(Side.CLIENT)
  val bakedModel = BakedModelFrame
  val frameTexture: ResourceLocation = new ResourceLocation("minecraft", "blocks/furnace_top")


  define {
    |#|:(0, 0, 0)(1, 16, 1) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(0, 0, 15)(1, 16, 16) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(15, 0, 0)(16, 16, 1) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(15, 0, 15)(16, 16, 16) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(1, 0, 0)(15, 1, 1) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(1, 15, 0)(15, 16, 1) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(0, 0, 1)(1, 1, 15) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(0, 15, 1)(1, 16, 15) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(15, 0, 1)(16, 1, 15) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(15, 15, 1)(16, 16, 15) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(1, 0, 15)(15, 1, 16) {
      |*(frameTexture, UV.auto(16))
    }
    |#|:(1, 15, 15)(15, 16, 16) {
      |*(frameTexture, UV.auto(16))
    }
  }
}

